src/pybamm/util.py

```diff
@@ -108,38 +109,100 @@ def __getitem__(self, key):
             ) from error
 
-    def search(self, key, print_values=False):
+    def _find_matches(self, search_key, known_keys):
         """
-        Search dictionary for keys containing 'key'. If print_values is True, then
-        both the keys and values will be printed. Otherwise just the values will
-        be printed. If no results are found, the best matches are printed.
+        Helper method to find exact and partial matches for a given search key.
+
+        Parameters
+        ----------
+        search_key : str
+            The term to search for in the keys.
+        known_keys : list of str
+            The list of known dictionary keys to search within.
+
+        """
+        exact = [key for key in known_keys if search_key in key.lower()]
+        partial = difflib.get_close_matches(search_key, known_keys, n=5, cutoff=0.5)
+        return exact, partial
+
+    def search(self, keys: str | list[str], print_values: bool = False):
+        """
+        Search dictionary for keys containing all terms in 'keys'.
+        If print_values is True, both the keys and values will be printed.
+        Otherwise, just the keys will be printed. If no results are found,
+        the best matches are printed.
+
+        Parameters
+        ----------
+        keys : str or list of str
+            Search term(s)
+        print_values : bool, optional
+            If True, print both keys and values. Otherwise, print only keys.
+            Default is False.
         """
-        key_in = key
-        key = key_in.lower()
 
-        # Sort the keys so results are stored in alphabetical order
-        keys = list(self.keys())
-        keys.sort()
-        results = {}
+        if not isinstance(keys, (str, list)):
+            raise TypeError(
+                f" 'keys' must be a string or a list of strings, got {type(keys)}"
+            )
 
-        # Check if any of the dict keys contain the key we are searching for
-        for k in keys:
-            if key in k.lower():
-                results[k] = self[k]
+        if isinstance(keys, str):
+            if not keys.strip():
+                raise ValueError(
+                    "The search term cannot be an empty or whitespace-only string."
+                )
+            original_keys = [keys]
+            search_keys = [keys.strip().lower()]
 
-        if results == {}:
-            # If no results, return best matches
-            best_matches = self.get_best_matches(key)
+        elif isinstance(keys, list):
+            if all(not str(k).strip() for k in keys):
+                raise ValueError(
+                    "The 'keys' list cannot contain only empty or whitespace strings."
+                )
+
+            if not all(isinstance(k, str) for k in keys):
+                raise TypeError("All elements in the 'keys' list must be strings.")
+
+            original_keys = keys
+            search_keys = [k.strip().lower() for k in keys if k.strip()]
```

Comment:
Why the `if` condition here?

Reply:
This condition was added to exclude empty strings or strings consisting only of whitespaces from `search_keys`, making the search more efficient and focused on the  meaningful terms only.

Reply:
Sorry, I should have added more details in -

You can either do -
```py
[k.strip().lower() for k in keys if k.strip()]
```
which skips empty/whitespace strings but also calls `strip()` in an `if` condition for each element of the list (but we save `lower()` calls as it is not called on empty/whitespace strings).


or -
```py
[k.strip().lower() for k in keys]
```
which calls `lower()` on every string (but we save calling `strip()` in an `if` condition for each element of the list).

I hope this makes sense. Both the expressions will give you the same result, but both the expressions call `strip()` and `lower()` on different number of elements. My questions is, which one of these is faster? If you have already experimented with the speed then I am okay with having the `if-strip()` implementation (could you please attach the results here), but if you haven't, then you should generate a large dummy string dataset (not necessarily pybamm oriented, maybe [Faker](https://faker.readthedocs.io/en/master/) can help) and [`timeit`](https://docs.python.org/3/library/timeit.html) both the implementations.

Remember - ["Our mental image of what code should be faster or slower is often wrong, or doesn't make much difference. The only way to really improve code performance is empirically, through measurements."](https://github-pages.ucl.ac.uk/rsd-engineeringcourse/ch08performance/015mandels.html)

I would have experimented with this but I am low on time for the next few weeks 😬 

Thank you again for taking this up!

Reply:
> My questions is, which one of these is faster?

I don't have a benchmark to corroborate either, but I am sure the second one, i.e.,

```python
[k.strip().lower() for k in keys]
```

would be faster – because the first one calls `strip()` twice for non-empty strings, once per condition and once per comprehension. For the second one, the `lower()` operation on empty/whitespace strings is very fast since there's nothing to convert.

Reply:
> My question is, which one of these is faster?

I don't have a benchmark to corroborate either, but I am sure the second one, i.e.,

```python
[k.strip().lower() for k in keys]
```

would be faster – because the first one calls `strip()` twice for non-empty strings, once per condition and once per comprehension. For the second one, the `lower()` operation on empty/whitespace strings is very fast since there's nothing to convert.

Reply:
That was my initial thought too 👀

Reply:
Thank you for discussing this ,I understand the trade-offs now. Though I haven't benchmarked the two implementations yet,  I will work on it and share the results so we can proceed accordingly.

Reply:
I worked on the benchmark 

- taking a set of 5000 random strings
- number of white spaces being a random number between 1 and 50.
- number of words in the keys also being random number between 1 and 50.
- shuffling the words and the whitespaces.
- getting the avergae result for 10,000 such cases

And this was the result:

```python
  [k.strip().lower() for k in keys if k.strip()] 
  ```

![Screenshot from 2024-12-19 19-15-11](https://github.com/user-attachments/assets/c195cd53-b5bd-431d-a9a6-090aff15d197)

![Screenshot from 2024-12-19 19-14-53](https://github.com/user-attachments/assets/490993af-31e2-4a32-890f-025e8a8b213f)

```python 
 [k.strip().lower() for k in keys] 
 ```

![Screenshot from 2024-12-19 19-15-03](https://github.com/user-attachments/assets/39b08f0c-0f95-45f0-8997-76205b3cb6a4)

![Screenshot from 2024-12-19 19-14-42](https://github.com/user-attachments/assets/60db737f-b924-44e4-b43a-6202f50021b4)

Reply:
Interesting! I seem to have the opposite result for random data containing a random sample of arbitrary-length strings of ASCII characters and non-empty whitespace strings. Here are the results from a script I wrote (10,000 iterations for each size):

```
Benchmark Results:
--------------------------------------------------
Method 1 is  [k.strip().lower() for k in keys if k.strip()]
Method 2 is  [k.strip().lower() for k in keys]
Size       Method 1 (ms)   Method 2 (ms)   Difference
--------------------------------------------------
100        0.006           0.004           Method 2 is 35.6% faster
1000       0.061           0.039           Method 2 is 56.1% faster
10000      0.591           0.442           Method 2 is 33.8% faster
100000     6.464           4.882           Method 2 is 32.4% faster
```

Reply:
In this case, the assumption is that all words in the dataset have whitespaces to strip, which explains the performance differences observed. However, in my benchmark, I considered a scenario where only a random subset (around 50%) of the words have whitespaces, while the rest are plain words without any whitespaces. I also included empty strings, which are filtered out by the condition in the code, as originally intended, before being included in the final search keys.

Reply:
Ah, I see. I tried it again by adding empty strings in my benchmarking script – so that it contains 20% empty strings, 20% strings with whitespaces, and 60% of them with potential padding, in no particular order for higher entropy, and the second one was still faster.

I would say that this experiment segues into a separate discussion, which is not in line with real-life usage of the functionality – as we can notice, both methods are quite fast (and the difference is unnoticeable) for search queries that are not too large (no one would search for a thousand terms at once).

What this boils down to is that while calling `lower()` is probably slow (and redundant) on empty strings, and we are raising an error on their inclusion in the search term, there is a possibility of non-empty strings in the search keys being imperfect, such as `"Electrolyte concentration "` (note the extra whitespace at the end), where stripping is arguably good to do so that our search terms are refined. So, I wouldn't worry about speed, and I agree with your assessment in https://github.com/pybamm-team/PyBaMM/pull/4650#discussion_r1891170142 – we can keep this as is.

Reply:
Thank you for the detailed insights! I agree that the performance difference in both cases is negligible, especially for practical use cases. Let’s proceed with the if condition, as it is now.